File: hash/almost_equal_pairs/almost_equal_pairs.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <deque>
using namespace std;

int countAlmostEqualPairs(vector<int> &nums)
{ // LeetCode Q.3267.
    // Each num's transformed nums set.
    unordered_map<int, unordered_set<int>> newNumsTable;
    unordered_map<int, double> oldNums2Counts, newNums2Counts;

    deque<int> queue;
    unordered_set<int> newNums;

    for (auto num : nums)
    {
        oldNums2Counts[num]++;

        if (newNumsTable.find(num) == newNumsTable.end())
        {
            queue.push_back(num);

            int rounds = 2;
            while (rounds > 0)
            {
                int transformations = queue.size();

                while (transformations > 0)
                {
                    int newNum = queue.front();
                    queue.pop_front();

                    string newNumStr = to_string(newNum);
                    int newNumLen = newNumStr.length();

                    for (int back_idx = 0; back_idx < newNumLen - 1; back_idx++)
                    {
                        for (int front_idx = back_idx + 1; front_idx < newNumLen; front_idx++)
                        {
                            swap(newNumStr[back_idx], newNumStr[front_idx]);

                            newNum = stoi(newNumStr);
                            if (newNum != num && newNums.find(newNum) == newNums.end())
                            {
                                newNums.insert(newNum); // Only take real new nums.
                                queue.push_back(newNum);
                            }

                            swap(newNumStr[back_idx], newNumStr[front_idx]);
                        }
                    }

                    transformations--;
                }

                rounds--;
            }

            newNumsTable[num] = newNums;
            queue.clear(); // Reset for future usage.
            newNums.clear();
        }

        int numLen = to_string(num).length();

        for (auto newNum : newNumsTable[num])
        {
            newNums2Counts[newNum]++;
            if (to_string(newNum).length() == numLen)
                newNums2Counts[newNum] -= 0.5;
        }
    }

    double almost_equal_pairs = 0.0;
    for (auto &pair : oldNums2Counts)
        almost_equal_pairs += pair.second * (pair.second - 1) / 2;

    for (auto num : nums)
        almost_equal_pairs += newNums2Counts[num];

    return int(almost_equal_pairs); // Return int form.
}

File: hash/almost_equal_pairs/almost_equal_pairs.py

def count_almost_equal_pairs(nums: list[int]) -> int:  # LeetCode Q.3267.
    new_nums_table: dict[int, set[int]] = dict()  # Each num's transformed nums set.

    old_nums2counts: dict[int, float] = dict()
    new_nums2counts: dict[int, float] = dict()

    queue: list[int] = []
    new_nums: set[int] = set()
    for num in nums:
        if num not in old_nums2counts.keys():
            old_nums2counts.update({num: 0})
        old_nums2counts[num] += 1

        if num not in new_nums_table.keys():
            queue.append(num)

            rounds = 2
            while rounds > 0:
                transformations = len(queue)
                while transformations > 0:
                    new_num = queue.pop(0)
                    digits = list(str(new_num))
                    new_num_len = len(digits)

                    for back_idx in range(new_num_len - 1):
                        for front_idx in range(back_idx + 1, new_num_len):
                            digits[back_idx], digits[front_idx] = digits[front_idx], digits[back_idx]

                            new_num = int("".join(digits))
                            if new_num != num and new_num not in new_nums:
                                new_nums.add(new_num)  # Only take real new nums.
                                queue.append(new_num)

                            digits[back_idx], digits[front_idx] = digits[front_idx], digits[back_idx]

                    transformations -= 1

                rounds -= 1

            new_nums_table.update({num: new_nums.copy()})  # Copy: prevent modification.
            queue.clear()
            new_nums.clear()

        num_len = len(str(num))
        for new_num in new_nums_table[num]:
            if new_num not in new_nums2counts.keys():
                new_nums2counts.update({new_num: 0})
            new_nums2counts[new_num] += 1

            if len(str(new_num)) == num_len:
                new_nums2counts[new_num] -= 0.5

    almost_equal_pairs = 0
    for _, count in old_nums2counts.items():
        almost_equal_pairs += count * (count - 1) // 2

    for num in nums:
        if num in new_nums2counts.keys():
            almost_equal_pairs += new_nums2counts[num]

    return int(almost_equal_pairs)  # Return int form.
